### DISCUSSION TITLE
Computing harmonic fields with a time domain simulation

### DISCUSSION THREAD

**Discussion by kalosu** [NanoComp/meep#2398] [Q&A]
Hello there again friends from meep community, 

This time I would like to ask for your help with another issue that I have.
I am interested in extracting the harmonic field produced by a circular Bragg grating structure which has an embedded point-like dipole emitter in between. For reference, you can take a look to the publication https://opg.optica.org/oe/fulltext.cfm?uri=oe-30-10-15913&id=471709&ibsearch=false. 

Also, you can take a look to the material configuration that I am using within MEEP: (I am using cylindrical coordinates)

![cell_air_cbg_w_gold_orig](https://user-images.githubusercontent.com/49510503/218786877-15ede492-9ee2-44c8-a0a4-adb854028101.png)

The black blocks found at the same height at where my dipole source is found correspond to the circular Bragg grating structure. 
Below the grating structure I have a thin SiO2 layer and below this layer I have a gold layer which extends also across the thickness of the Absorber region that I have defined for the bottom boundary along z (for the rest of boundaries I am using standard PMLs)

Finally, the medium on top and in between the grating structure corresponds to air and therefore has an index value of 1.

On the reference publication that I have shared, the guy working on the simulations used a FEM solver (JCMwave) in the frequency domain for solving for the harmonic fields produced by the complete structure (dipole + grating + sio2 layer, etc).

I have been trying to replicate his results using MEEP for which so far I have replicated the results for the related extraction efficiency. In my case, I calculated the extraction efficiency by using the total power from the LDOS calculation and by using a near to far field calculation for points defined over a circle section that intersects with my Meep simulation domain(r from 0 to rmax and z from -z to z).
This evaluation agrees nicely to the results obtained through the frequency domain FEM simulation.

Following, I tried to move to the part in where I extract the field at a given distance from the grating structure. In my case, I use a DFT monitor for computing the cylindrical field components at a distance of 2 microns from the grating structure's top interface.

As you can observe in the following graph, I have not been able to obtain the same field distribution as from the FEM simulation
![cbg_field_comp_meep_jcmwave_z_2_res_80_meep_w_gold](https://user-images.githubusercontent.com/49510503/218791599-13824ed9-02d0-4b69-875c-a834cbe0a535.png)

The field generated by such a structure should have a smooth and clean Gaussian like shape as observed from the FEM curves. Nevertheless, as you can see, the results that I obtain using MEEP are somehow more "noisy" or contain more high frequency information.

I suspect that this is due to the fact that I am using a pulsed time profile for my dipole source. I have tried the following things without really having any significant improvement for the evaluated field components:

1. I tried using the frequency domain solver. Nevertheless, by setting the conductivity of gold manually my simulation just did not converge. I got an error indicating that there were some convergence issues.
2. I have tried increasing the length of the simulation runtime by reducing the bandwidth of my pulse (from 0.08 to 0.01) and by also reducing the value of my decay_by parameter for the simulation stop criteria (from 10^-11 to 10^-15) . Nevertheless, this did not result in any difference.
3. I have tried increasing the resolution and the PML thickness but again, this has not helped.

What could be the causes for the observed field structure and its difference with respect to the FEM results? Do you have any suggestion about something that I could try or something that I would need to check for? 

As always, thanks for your feedback.

**Comment by smartalecH** [comment]
So computing frequency-domain quantities using time-domain methods can be particularly challenging if the problem is highly resonant (which yours is). This is because in order to accurately resolve the Fourier fields, you need to ensure they sufficiently _converge_, which means you need to wait until the resonance "rings out". 

If you stop too early, you effectively "window" your time-domain fields with a rect, which _convolves_ your frequency-domain fields with a sinc. From your plot, I definitely see some sinc-like characteristics. 

Now _one_ easy fix would be to just let your simulation run longer and longer and you should notice the sidelobes gradually shrinking. (Which you tried, to no avail). 

Alternatively, you could _exploit_ the fact that we know the problem is highly resonant, and build a Pade approximate from your time series. This effectively allows you to overcome the Fourier uncertainty limitation (since you have a-priori knowledge of the problem) and you can compute very accurate frequency responses with relatively short time sequences (a few periods or so). The tradeoff here is _storage_ (and obviously implementation complexity) as you now have to store data at every point in space you care about _and_ at every timestep. 

For an early take on how to do this, check out #2192. (cc @Arjun-Khurana for a status update). 

Lastly, I'll mention that we are finding some small bugs in the cylindrical coordinates solver, particularly with sources right at r=0. But I think we recently pushed fixes for those. (cc @oskooi for any further comments in that regard).

**Reply by smartalecH** [reply]
Ah looking at your plot, I see the x axis is position, not frequency. 

Regardless, the idea of windowing is still a factor here, just in space/reciprocal space, rather than time/frequency space. 

The Pade approximate should still help too.

**Reply by smartalecH** [reply]
Since the grating is finite, I would almost expect some sinc-like behavior in the spatial fields... but I could be wrong. 

If you gradually make the grating longer and longer, do the sidelobes shrink?

**Comment by oskooi** [comment]
> I have tried increasing the length of the simulation runtime by reducing the bandwidth of my pulse (from 0.08 to 0.01) and by also reducing the value of my decay_by parameter for the simulation stop criteria (from 10^-11 to 10^-15) . Nevertheless, this did not result in any difference.

Have you tried increasing the `cutoff` parameter of the `GaussianSource`? Try doubling it to 10 from its default value of 5.

For some context, we are currently investigating a potential bug in the $z$-PML at $r=0$ in cylindrical coordinates (#2182, #2383). The workaround for now seems to involve using a pulsed source with narrow bandwidth *and* long cutoff. This is documented in the tutorial [Extraction Efficiency of a Light-Emitting Diode (LED)](https://meep.readthedocs.io/en/latest/Python_Tutorials/Local_Density_of_States/#extraction-efficiency-of-a-light-emitting-diode-led): "In this example, the cutoff property of the GaussianSource is doubled to 10 from its default value of 5. The spectral bandwidth (fwidth) is 5% of the center frequency of 1.0."

**Comment by kalosu** [comment]
Hi everyone, thanks for the suggestions.

I gave it another try to the implementation in cylindrical coordinates by using the following parameters:
Using a pulsed gaussian source defined with:
fcen = 1/0.930 (wavelength: 930 nm)
fwidth = fcen*0.08 
cutoff = 20

Then as for the simulation run time I used the stop criteria "stop_when_fields_decayed" with a decay_by parameter value of 1e-16 by monitoring the field value at the position of the dipole emitter. 

Using this configuration, I obtained the following results:
![cbg_field_comp_meep_jcmwave_z_2_cyl_bw_8per_cutoff_20_decay_by_min16](https://user-images.githubusercontent.com/49510503/219310691-3bc50fbb-be2a-469b-a2d7-5ead1e3f8f8a.png)

You can clearly see now how the results agree in better way to the ones from the FEM simulation.

I suppose that if I run the simulation for an even longer time (by again adjusting the decay_by value and increasing the cutoff??) I would be able to obtain better results.

As for the field phase, I can also show you the comparison: 
![cbg_field_comp_meep_jcmwave_z_2_cyl_bw_8per_cutoff_20_decay_by_min16_phase](https://user-images.githubusercontent.com/49510503/219311666-313d814b-6a2b-43ca-b78d-7552e3ca3052.png)

I wanted to give it another try to the frequency domain solver. I have the feeling that I should be able to compute the frequency domain response but I am not really sure if I am setting the parameters in a proper manner.

As you know and as mentioned on Meep's documentation for the frequency domain solver; Lorentz-Drude dispersive materials are not supported by the frequency domain solver and one way to overcome this is by specifying the conductivity of the used material at the desired frequency.

As you have already observed, I am using gold as a back reflective layer beneath my grating structure. If I want to use the frequency domain solver for such a material configuration, how should I set up the conductivity of gold in this case?

At 930 nm gold as a permittivity value equal to: -35.328 + i2.2675.
In order to set the properties of gold for the frequency domain solver,  I tried setting the conductivity value as mentioned in the documentation. 
The question then for me comes: What should be the value for $\epsilon_{\infty}$ that I need to use? Should it be 1? Should it be the value at a shorter wavelength for which epsilon is positive? What about the imaginary part, should I use the imaginary part from the permittivity at 930nm? At a different wavelength?

If you could guide me with how to properly set these parameters I think I could then use the frequency domain solver.

**Comment by kalosu** [comment]
Hi there again,

I was trying to use the cylindrical simulations in combination with the frequency domain solver. For this case, I could not get the results to converge,i.e, the shown residue values blows up indefinitely after some run time.

I then decided to give it a try with the case of the dipole emitter sitting at r,z =0 (er dipole) in an open homogeneous space. For this case I also experienced the same issue, after a few iteration steps the residue blows up. (For this case I used a PML for every boundary).

What could be the reason for this? Is the frequency domain solver also supported for cylindrical coordinate simulations?

Do I need to adjust any parameter for the defined ContinuousSource used as an excitation?

**Comment by kalosu** [comment]
At the end I have tried the following:
1. Using the frequency domain solver in cylindrical coordinates: As described above, I could not get the solver to converge even for the dipole in open homogeneous space case.
2. Using the frequency domain solver in cartesian coordinates in 2D: For this case, the frequency domain solver worked even with a gold layer included for which I set the conductivity by setting epsilon to 1 with the imaginary part of epsilon by the value that I mentioned on my previous post (2.675). You can take a look to the results:
![meep_freq_domain_cbg_yes_gold_sio2_extended_pml](https://user-images.githubusercontent.com/49510503/219444176-5b9f0041-2286-4934-9d1b-413a3b53524d.png)
![meep_freq_domain_cbg_yes_gold_sio2_extended_pml_phase](https://user-images.githubusercontent.com/49510503/219444193-bc1209f2-ebb9-4a2c-ad7f-d2e6155eb7e3.png)

Above plot shows the amplitude and phase obtained for ex at a distance of 2 microns above the 2D grating structure defined in the xy plane. The difference between both curves is that for one I have included the thin gold layer and for the other I did not. You can see that the obtained results are somehow in agreement. Nevertheless, this 2D simulation is not equivalent to my 3D full configuration for which the grating structure has cylindrical symmetry, in this case the assumption is that the grating is infinitely extended along z right? I can imagine this is one of the reasons why the computed field does not have the nice Gaussian like distribution obtained from the FEM simulation.

3. I then decided to try using the frequency domain solver but in 3D. Nevertheless, as you might expect this simulation just takes forever. In fact, I could not even get the simulation to start.

Any thoughts on what else could I do in order to compute the steady state field or the harmonic field response? At the end I am trying to use the cylindrical coordinates simulation by using a narrow bandwidth of 0.03 times the central frequency, a cutoff of 25 and as a stop criteria I am using the field_decay_by settings at the z position at where I want to compute my field (with a value of 1e-17)

**Comment by kalosu** [comment]
Hello again meep community, 

I was finally able to compute the "steady-state" or harmonic-field like component that I was interested in. As I had mentioned before, such a structure produces a close to ideal Gaussian like field distribution in the far field. In the publication that I have referred to, they have shown that it is possible to obtain an extraction efficiency from the dipolar source for a large value of above 90% for a relatively low angular extension, roughly 20 degrees in half angle.

As I mentioned before as well, I have been able to replicate the extraction efficiency curve shown in the paper using my Meep implementation in cylindrical coordinates. You can take a look to it: 
![ext_eff_cbg_filled_air_930_sa](https://user-images.githubusercontent.com/49510503/220688882-d9754b1e-d42e-4de6-80cb-1ab237232989.png)

As shown above, most of the power is redirected to the upper half space as shown by the value of the Poynting flux component along z. 

What I then needed to do was to change the refractive index for the medium filling the trenches in the cavity structure. On the paper, they have used air and in my case, I am interested in knowing how does the field and the extraction efficiency change for such a structure once the trenches medium's index is changed. For that, I have taken the value of 1.5025 (value for a polymer material that we use). You can see how does the refractive index distribution changes now as
![cell_air_cbg_n_gold_cyl_n2f](https://user-images.githubusercontent.com/49510503/220690358-a61c3fcf-5b88-4777-bef6-b358dafec45d.png)

Notice that beneath the grating structure I am using SiO2 which at the evaluation wavelength (930nm) has an index value of 1.45, therefore the lower index contrast.

My issue then comes when I try to re compute the extraction efficiency. As mentioned in my original post, I compute this by taking the ratio of the power calculated in the far field as a function of a given maximum "observation angle" to the power emitted by the source element. For the dipole total power, I was trying to use the LDOS based calculation which is also the value that I used for the air based evaluations. 

Now, from the following curve you can directly see that there is something wrong this time for the evaluation:

![fluxes_cbg_filled_ips_with_gold_930_sa](https://user-images.githubusercontent.com/49510503/220702315-62416975-e8c8-43b3-a3e7-f25075438682.png)

1. From above curve it can be seen that the total power computed by the LDOS approach is smaller than the power calculated by the far field monitors and by the total power calculated from using flux monitors at the upper and right boundaries (since at the bottom I have gold, I have not used a monitor here; ideally most of the power should be reflected by the gold layer). 
2. Additionally, the LDOS power is negative: (-0.001351)

What can be the reason behind this issue with the total power calculated by using the LDOS? Do you have any suggestion on what could I do? Would using flux monitors around the dipole source element work as an alternative? Would I get the same accuracy without having to increase the resolution of my grid?

**Comment by kalosu** [comment]
Hello again,

So I was giving it another look to the simulation results. I decided to run a resolution sweep to see if somehow the results for the LDOS based estimation would converge as a function of resolution.

You can take a look to the evaluations I did using the polymer material and air for filling the trenches:

![cbg_filled_air_with_gold_930_sa_res_sweep](https://user-images.githubusercontent.com/49510503/220904838-4264b984-70c2-4bed-88a8-0948b0cca70e.png)


![cbg_filled_ips_with_gold_930_sa_res_sweep](https://user-images.githubusercontent.com/49510503/220904950-1d1d862d-c1b8-45c8-a36b-4d241e6f46e4.png)

As you can see, for the flux monitor based estimations the values somehow tend to stay within well defined boundaries. For the LDOS based case nevertheless, you can see how the computed extraction efficiency fluctuates and the calculated efficiencies are not even realistic.

Again, what could be the reason behind this? Any thoughts?
